geoportailv3/static/js/layerpermalinkservice.js

```diff
@@ -0,0 +1,254 @@
+/**
+ * @fileoverview This files defines an Angular Service for managing the
+ * selected layers permalinks
+ */
+
+goog.provide('app.LayerPermalinkManager');
+
+goog.require('app');
+goog.require('app.GetLayerForCatalogNode');
+goog.require('app.StateManager');
+goog.require('app.Themes');
+goog.require('ngeo.BackgroundLayerMgr');
+
+
+
+/**
+ * @constructor
+ * @param {app.StateManager} appStateManager
+ * @param {ngeo.BackgroundLayerMgr} ngeoBackgroundLayerMgr
+ * @param {app.GetLayerForCatalogNode} appGetLayerForCatalogNode
+ * @param {app.Themes} appThemes
+ * @ngInject
+ */
+app.LayerPermalinkManager =
+    function(appStateManager, ngeoBackgroundLayerMgr,
+        appGetLayerForCatalogNode, appThemes) {
+  /**
+   * @type {app.Themes}
+   * @private
+   */
+  this.appThemes_ = appThemes;
+
+  /**
+   * @type {app.StateManager}
+   * @private
+   */
+  this.stateManager_ = appStateManager;
+
+  /**
+   * @type {ngeo.BackgroundLayerMgr}
+   * @private
+   */
+  this.backgroundLayerMgr_ = ngeoBackgroundLayerMgr;
+
+  /**
+   * @type {app.GetLayerForCatalogNode}
+   * @private
+   */
+  this.getLayerFunc_ = appGetLayerForCatalogNode;
+};
+
+
+/**
+ * @param {Array.<ol.layer.Layer>} layers
+ * @private
+ */
+app.LayerPermalinkManager.prototype.setLayerState_ = function(layers) {
+  var layerIds = goog.array.map(layers, function(layer) {
+    return layer.get('queryable_id');
+  });
+  var opacities = goog.array.map(layers, function(layer) {
+    return layer.getOpacity();
+  });
+  var bgLayer = this.backgroundLayerMgr_.get(this.map_);
+  this.stateManager_.updateState({
+    'layers': layerIds.join('-'),
+    'opacities': opacities.join('-'),
+    'bgLayer': bgLayer.get('label')
+  });
+};
+
+
+/**
+ * @param {Array.<number>} layerIds
+ * @param {Array.<Object>} themes
+ * @private
+ */
+app.LayerPermalinkManager.prototype.applyLayerStateToMap_ =
+    function(layerIds, themes) {
+  var flatCatalogue = [];
+  for (var i = 0; i < themes.length; i++) {
+    var theme = themes[i];
+    goog.array.extend(flatCatalogue,
+        app.LayerPermalinkManager.getAllChildren_(theme.children)
+    );
+  }
+  goog.array.forEach(layerIds.reverse(),
+      function(layerId, layerIndex) {
+        var node = goog.array.find(flatCatalogue, function(catItem) {
+          return catItem.id === layerId;
+        });
+        var layer = this.getLayerFunc_(node);
+        // set opacity trough metadata to not interfere
+        // with the layer opacity manager service
+        layer.get('metadata')['start_opacity'] =
+            this.getStateValue_('opacities').reverse()[layerIndex];
```

Comment:
Also, it is strange to call `reverse()` on something that may be `undefined`. (The return type of `getStateValue_` is `Array.<number>|undefined`.)